File: leetcode/Python/search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r, p = 0, len(nums)-1, 0
        for i in range(1,len(nums)):
            if nums[i]<nums[p]:
                p+=1
                break
            p+=1
        if nums[p]<=target<=nums[-1]:
            l=p
        else:
            r=p-1
        while l<=r:
            p=(l+r)//2
            if nums[p]<target:
                l=p+1
            elif nums[p]>target:
                r=p-1
            else:
                return p
        return -1

File: leetcode/C++/search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int p=0,l=0,r=nums.size()-1;
        for(int i=1; i<nums.size(); ++i) if(nums[p++]>nums[i]) break;
        if(nums[p]<=target && target<=nums.back()) l=p; 
        else r=p-1;
        while(l<=r){
            p=(l+r)/2;
            if(nums[p]<target) l=p+1;
            else if(nums[p]>target) r=p-1;
            else return p;
        }
        return -1;
    }
};